java/gradle.htmlui/src/org/netbeans/modules/gradle/htmlui/GradleArchetype.java

```diff
@@ -18,59 +18,74 @@
+        projectDirs.add(rootDir);
+        
         Enumeration<? extends FileObject> en = templates.getChildren(true);
         while (en.hasMoreElements()) {
             FileObject template = en.nextElement();
-            if (!template.isData()) {
-                continue;
-            }
-            String relativeParent = FileUtil.getRelativePath(templates, template.getParent());
-            Object packageAttr = template.getAttribute("package"); // NOI18N
-            if (packageAttr instanceof String) {
-                String packageName = mf.format(packageAttr).replace('.', '/');
-                relativeParent += "/" + packageName;
-            }
-            FileObject destinationFolder = FileUtil.createFolder(projectFo, relativeParent);
+            String relativePath = FileUtil.getRelativePath(templates, template);
+            if (template.isFolder()) {
+                File dir = new File(rootDir, relativePath);
+                ops.createFolder(dir);
+                Object projectAttr = template.getAttribute("project"); // NOI18N
+                if (Boolean.TRUE == projectAttr) {
+                    projectDirs.add(dir);
+                }
+            } else if (template.isData()) {
+                Object packageAttr = template.getAttribute("package"); // NOI18N
+                if (packageAttr instanceof String) {
+                    String relativeParent = FileUtil.getRelativePath(templates, template.getParent());
+                    String packageName = mf.format(packageAttr);
+                    File sourceRoot = new File(rootDir, relativeParent);
+                    ops.createPackage(sourceRoot, packageName);
+                    File packageDir = new File(sourceRoot, packageName.replace('.', '/'));
 
-            FileObject previous = destinationFolder.getFileObject(template.getNameExt());
-            if (previous != null) {
-                previous.delete();
+                    Map<String, Object> pparams = new HashMap<>(params);
+                    pparams.put("package", packageName); //NOI18N
+                    copyDataTemplate(ops, template, new File(packageDir, template.getNameExt()), pparams);
+                } else {
+                    copyDataTemplate(ops, template, new File(rootDir, relativePath), params);
+                }
             }
+        }
+        for (File projectDir : projectDirs) {
+            ops.addProjectPreload(projectDir);
+        }
+    }
 
-            FileBuilder fb = new FileBuilder(template, destinationFolder);
-            fb.withParameters(params);
-            fb.defaultMode(Mode.COPY);
-
-            FileObject copied = fb.build().iterator().next();
-
-            assert copied != null && copied.getNameExt().equals(template.getNameExt()) : "Created " + copied;
+    private static void copyDataTemplate(TemplateOperation ops, FileObject template, File target, Map<String, Object> params) {
+        Object importantAttr = template.getAttribute("important");
```

Comment:
The `important` attribute enhances the API, but I am not sure if the previous API was documented anywhere.

Reply:
Yes that's true, though this module is not providing any public API/SPI at the moment. Maybe some next iterations when the GradleArtifact would be made available for reuse. Infortunately I have at least two branches open to improve the Gradle Wizards, but I have not got the chance to bring them in a presentable state.

Reply:
When you think of improving the wizard, there are two things I'd suggest to address the infamous Gradle problems:
* fix the Gradle version - e.g. always generate `gradlew` - Gradle isn't trying to keep any backward compatibility. The only way to create a Gradle project and make it work after one, two, three years is to fix Gradle version with `gradlew`.
* the wizard should select (a recommended) JDK - Gradle is well known to not work with newest JDKs - the wizard should push user strongly to provide a JDK which is known to work for the project and/or selected Gradle version

Unrelated to this issue, of course. Just my 2 Kč experience of making Geertjan's [java frontend application](https://github.com/JaroslavTulach/oraclejet4j) working again after twelve (only!) months.

Reply:
The JDK selection is already on my branch.
I can set the wrapper on by default right now, that would be a trivial change and improve the compatibility especially as Gradle 7.0 is already on the design desk.
It seems I have to fix the tests as well.